hack/e2e-suite/guestbook.sh

```diff
@@ -33,25 +33,75 @@ export KUBECTL KUBE_CONFIG_FILE
 source "${KUBE_ROOT}/cluster/kube-env.sh"
 source "${KUBE_VERSION_ROOT}/cluster/${KUBERNETES_PROVIDER}/util.sh"
 
+function wait_for_running() {
+  echo "Waiting for pods to come up."
+  frontends=$(${KUBECTL} get pods -l name=frontend -o template '--template={{range.items}}{{.id}} {{end}}')
+  master=$(${KUBECTL} get pods -l name=redis-master -o template '--template={{range.items}}{{.id}} {{end}}')
+  slaves=$(${KUBECTL} get pods -l name=redisslave -o template '--template={{range.items}}{{.id}} {{end}}')
+  pods=$(echo $frontends $master $slaves)
+
+  all_running=0
+  for i in $(seq 1 30); do
+    sleep 10
+    all_running=1
+    for pod in $pods; do
+      status=$(${KUBECTL} get pods $pod -o template '--template={{.currentState.status}}') || true
+      if [[ "$status" == "Pending" ]]; then
+        all_running=0
+        break
+      fi
+    done
+    if [[ "${all_running}" == 1 ]]; then
+      break
+    fi
+  done
+  if [[ "${all_running}" == 0 ]]; then
+    echo "Pods did not come up in time"
+    exit 1
+  fi
+}
+
+function teardown() {
+  ${KUBECTL} stop -f "${GUESTBOOK}" 
+}
+
 prepare-e2e
 
 GUESTBOOK="${KUBE_ROOT}/examples/guestbook"
 
-echo "WARNING: this test is a no op that only attempts to launch guestbook resources."
 # Launch the guestbook example
 ${KUBECTL} create -f "${GUESTBOOK}"
 
-sleep 15
-
-POD_LIST_1=$(${KUBECTL} get pods -o template '--template={{range.items}}{{.id}} {{end}}')
-echo "Pods running: ${POD_LIST_1}"
-
-# TODO make this an actual test. Open up a firewall and use curl to post and
-# read a message via the frontend
-
-${KUBECTL} stop -f "${GUESTBOOK}"
-
-POD_LIST_2=$(${KUBECTL} get pods -o template '--template={{range.items}}{{.id}} {{end}}')
-echo "Pods running after shutdown: ${POD_LIST_2}"
+trap "teardown" EXIT
+
+# Verify that all pods are running
+wait_for_running
+
+# Add a new entry to the guestbook and verify that it was remembered 
+FRONTEND_ADDR=$(${KUBECTL} get service frontend -o template '--template={{range .publicIPs}}{{.}}{{end}}:{{.port}}')
+MINION=$(${KUBECTL} get nodes -o template '--template={{range.items}}{{.id}} {{end}}' | cut -d " " -f 1 | cut -d "." -f 1)
+echo "Waiting for frontend to serve content"
+serving=0
+for i in $(seq 1 12); do
+  ENTRY=$(gcloud compute ssh $MINION --command "curl ${FRONTEND_ADDR}/index.php?cmd=get\&key=messages")
```

Comment:
This, and other uses of gcloud, have to change. This is a provider agnostic test. Can we not use a service proxy endpoint from the master to hit this, like how grafana is hit in https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/gce/util.sh?